HotFix/Utilities/Parser.cs

```diff
@@ -235,5 +235,223 @@ public static DateTime GetDateTime(this byte[] source, int offset, int count)
+                value /= 10;
+            }
+
+            if (negative) buffer[position++] = (byte)'-';
+
+            for (var i = 0; i < (position - initial) / 2; i++)
+            {
+                var temp = buffer[initial + i];
+                buffer[initial + i] = buffer[position - i - 1];
+                buffer[position - i - 1] = temp;
+            }
+
+            return position - initial;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int WriteFloat(this byte[] buffer, int position, double value)
+        {
+            var start = position;
+            var integer = (long) value;
+            var fraction = (long)Math.Round(Math.Abs(value) * 1000000L) % 1000000L;
+
+            position += buffer.WriteLong(position, integer);
+            buffer[position++] = (byte)'.';
+            position += buffer.WriteLong(position, fraction);
+
+            return position - start;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int WriteString(this byte[] buffer, int position, string value)
+        {
+            for (var i = 0; i < value.Length; i++)
+            {
+                buffer[position + i] = (byte)value[i];
+            }
+
+            return value.Length;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int WriteDateTime(this byte[] buffer, int position, DateTime value)
+        {
+            var year = value.Year;
+            buffer[position] = (byte)('0' + year / 1000);
+            year %= 1000;
+            buffer[position + 1] = (byte)('0' + year / 100);
+            year %= 100;
+            buffer[position + 2] = (byte)('0' + year / 10);
+            year %= 10;
+            buffer[position + 3] = (byte)('0' + year);
+
+            var month = value.Month;
+            buffer[position + 4] = (byte)('0' + month / 10);
+            month %= 10;
+            buffer[position + 5] = (byte)('0' + month);
+
+            var day = value.Day;
+            buffer[position + 6] = (byte)('0' + day / 10);
+            day %= 10;
+            buffer[position + 7] = (byte)('0' + day);
+
+            buffer[position + 8] = (byte)'-';
+
+            var hour = value.Hour;
+            buffer[position + 9] = (byte)('0' + hour / 10);
+            hour %= 10;
+            buffer[position + 10] = (byte)('0' + hour);
+
+            buffer[position + 11] = (byte)':';
+
+            var minute = value.Minute;
+            buffer[position + 12] = (byte)('0' + minute / 10);
+            minute %= 10;
+            buffer[position + 13] = (byte)('0' + minute);
+
+            buffer[position + 14] = (byte)':';
+
+            var second = value.Second;
+            buffer[position + 15] = (byte)('0' + second / 10);
+            second %= 10;
+            buffer[position + 16] = (byte)('0' + second);
+
+            var millis = value.Millisecond;
+
+            if (millis == 0) return 17;
```

Comment:
Not entirely sure this is valid - I though that in the `FIX Protocol` including milliseconds was a `type/setup decision` ie a decision made by both parties upfront? If so, then this could cause parsing errors on the other end...

We could make this decision a configuration aspect in the library and an optional parameter to this extension method. If that impacts performance too much, then let's just always write milliseconds?